### PR TITLE
fix Issue 16307 - [ndslice] should be safe

### PR DESCRIPTION
**Pull request by 9il**

### CODE REVIEW

File: std/experimental/ndslice/iteration.d [hunk 1]
```diff
@@ -104,6 +104,8 @@ import std.meta;
 import std.experimental.ndslice.internal;
 import std.experimental.ndslice.slice; //: Slice;
 
+@trusted:
```
Comment (MetaLang) [inline std/experimental/ndslice/iteration.d]: Having `@trusted` at the top of the file is a very bad idea. It runs a high risk of affecting things that you didn't intend to be `@trusted`.

File: std/experimental/ndslice/selection.d [hunk 2]
```diff
@@ -59,6 +59,8 @@ import std.meta; //: allSatisfy;
 import std.experimental.ndslice.internal;
 import std.experimental.ndslice.slice; //: Slice;
 
+@trusted:
```
Comment (MetaLang) [inline std/experimental/ndslice/selection.d]: Ditto.

File: std/experimental/ndslice/slice.d [hunk 3]
```diff
@@ -1093,6 +1099,7 @@ struct Slice(size_t _N, _Range)
                      && (isPointer!_Range || is(typeof(_Range.init[size_t.init]))))
                     || is(_Range == Slice!(N1, Range1), size_t N1, Range1)))
 {
+    @trusted:
```
Comment (MetaLang) [inline std/experimental/ndslice/slice.d]: Ditto.

File: std/experimental/ndslice/iteration.d (hunk 1 above)
Comment (9il) [inline std/experimental/ndslice/iteration.d]: All functions in `iterations` should be safe AFAIK.

File: std/experimental/ndslice/selection.d (hunk 2 above)
Comment (9il) [inline std/experimental/ndslice/selection.d]: All functions in `selection` should be safe AFAIK.

File: std/experimental/ndslice/slice.d (hunk 3 above)
Comment (9il) [inline std/experimental/ndslice/slice.d]: All methods in `Slice` should be safe AFAIK, except the methods, which manage underlaying data directly. They are marked `@system` explicitly

File: std/experimental/ndslice/slice.d [hunk 4]
```diff
@@ -1181,7 +1188,7 @@ struct Slice(size_t _N, _Range)
     }
 
     static if (!hasPtrBehavior!PureRange)
-    this(in size_t[PureN] lengths, in sizediff_t[PureN] strides, PtrShell!PureRange shell)
+    this(in size_t[PureN] lengths, in sizediff_t[PureN] strides, PtrShell!PureRange shell) @system
```
Comment (MetaLang) [inline std/experimental/ndslice/slice.d]: I'm not 100% sure but I *think* that as members of a template struct, `@system` should be inferred here. Regardless, it's the default anyway.

File: std/experimental/ndslice/slice.d [hunk 5]
```diff
@@ -1199,7 +1206,7 @@ struct Slice(size_t _N, _Range)
         strides = strides
         range = range or pointer to iterate on
     +/
-    this(in size_t[PureN] lengths, in sizediff_t[PureN] strides, PureRange range)
+    this(in size_t[PureN] lengths, in sizediff_t[PureN] strides, PureRange range) @system
```
Comment (MetaLang) [inline std/experimental/ndslice/slice.d]: Same here.

File: std/experimental/ndslice/slice.d [hunk 6]
```diff
@@ -1249,7 +1256,7 @@ struct Slice(size_t _N, _Range)
         if and only if all strides are positive.
     +/
     static if (is(PureRange == Range))
-    auto ptr() @property
+    auto ptr() @property @system
```
Comment (MetaLang) [inline std/experimental/ndslice/slice.d]: And here.

File: std/experimental/ndslice/iteration.d (hunk 1 above)
Comment (wilzbach) [inline std/experimental/ndslice/iteration.d]: @MetaLang's point was that it's a bad practice to annotate an entire module with `@trusted` as it can be very easy that something not safe slips through or will slip through in the future.

Comment (MetaLang) [inline std/experimental/ndslice/iteration.d]: What if something un-`@safe` is added in the future? It'll automatically be marked `@trusted`.

File: std/experimental/ndslice/selection.d (hunk 2 above)
Comment (MetaLang) [inline std/experimental/ndslice/selection.d]: See my previous answer.

File: std/experimental/ndslice/slice.d (hunk 4 above)
Comment (9il) [inline std/experimental/ndslice/slice.d]: No, the entire struct is `@trusted`

File: std/experimental/ndslice/slice.d (hunk 5 above)
Comment (9il) [inline std/experimental/ndslice/slice.d]: the entire struct is `@trusted`

File: std/experimental/ndslice/slice.d (hunk 6 above)
Comment (9il) [inline std/experimental/ndslice/slice.d]: the entire struct is `@trusted`

File: std/experimental/ndslice/slice.d (hunk 4 above)
Comment (MetaLang) [inline std/experimental/ndslice/slice.d]: IMO that's even worse.

**Comment by MetaLang** [comment]
I don't have the power to merge or reject PRs but I would personally not accept a PR that marked large swathes of code like this `@trusted`.

File: std/experimental/ndslice/iteration.d (hunk 1 above)
Comment (9il) [inline std/experimental/ndslice/iteration.d]: `iteration` module must manage only save operations on strides and lengths. It must not contain any other stuff.

**Comment by 9il** [comment]
`.slice.Slice` trusted, except only one methods that are designed for interaction with other libraries such as NumPy. They are `ptr` and `this`, all other methods must be save or must not be added into the structure.
`.iteration` & `.selection` - safe transformations on lengths and strides by definition. They must not manage pointers or underlaying ranges. The concept of ndslice and Mir library is to split system code from math code.

**Comment by schveiguy** [comment]
As strong a veto as I can have, we cannot accept `@trusted:` anywhere in any file.

The proper mechanism for templates is to mark the parts `@trusted` that should be with escapes, and let the compiler infer the `@safe` tag on the function itself. Mark your unittests as `@safe` to ensure they are being correctly inferred.

File: std/experimental/ndslice/slice.d [hunk 7]
```diff
@@ -565,6 +567,7 @@ slice(T,
 }
 
 /// ditto
+@trusted
```
Comment (aG0aep6G) [inline std/experimental/ndslice/slice.d]: This one is no good. T may have an unsafe postblit.

File: std/experimental/ndslice/slice.d [hunk 8]
```diff
@@ -585,6 +588,7 @@ auto slice(T,
...
```
Comment (aG0aep6G) [inline std/experimental/ndslice/slice.d]: Unsafe postblit breaks this one, too.

File: std/experimental/ndslice/slice.d [hunk 9]
```diff
@@ -727,6 +731,7 @@ Params:
 Returns:
     multidimensional D array
 +/
+@trusted
```
Comment (aG0aep6G) [inline std/experimental/ndslice/slice.d]: Also broken by unsafe postblit.

File: std/experimental/ndslice/slice.d [hunk 10]
```diff
@@ -805,6 +810,7 @@ Returns:
 Throws:
     $(LREF SliceException) if the array is not an n-dimensional parallelotope.
 +/
+@trusted
```
Comment (aG0aep6G) [inline std/experimental/ndslice/slice.d]: What's this needed for? As far as I see, the most unsafe thing here is a void initialization. And that's actually accepted in `@safe`.

**Comment by quickfur** [comment]
Yeah, I'm very uncomfortable with marking large swathes of code as `@trusted`. There is just too much room for error.  Generally, `@trusted` should be avoided as much as possible, and confined to as small a fragment of code as possible, so that it's easier to review and verify its trustworthiness.

Putting `@safe:` at the top of the module is OK, but definitely not `@trusted:`. Perhaps one approach is to put `@safe:` at the top and then narrow down exactly which parts of the code fail to compile, and either fix them or add appropriate `@trusted` helpers.

I'd also say that marking an entire function as `@trusted` is also suspect, unless the function is small. Marking template functions as `@trusted` is *especially* suspect, because the template arguments may contain `@system` primitives from user code that's beyond your control, and it would be very bad to have `@trusted` apply to such arbitrary code.

If a function needs to do `@system` operations but you're sure it can be trusted, the piece of code that performs said operations should be wrapped in a `@trusted` lambda, and we need to be absolutely sure that this code will never invoke anything outside its control, like a user-supplied type that may contain arbitrary code in `@system` methods.

File: std/experimental/ndslice/iteration.d (hunk 1 above)
Comment (quickfur) [inline std/experimental/ndslice/iteration.d]: Then it should be marked `@safe`. Not `@trusted`.

**Comment by schveiguy** [comment]
> I'd also say that marking an entire function as `@trusted` is also suspect, unless the function is small.

This I don't necessarily agree with. One trap to be careful to avoid is to do operations in `@trusted` escapes that affect data *outside* the trusted escape in such a way that the entire function becomes *essentially* trusted, but not marked as such. This is very dangerous as well.

File: std/experimental/ndslice/slice.d [hunk 11]
```diff
@@ -46,7 +46,7 @@ Returns:
 auto sliced(
     Flag!"replaceArrayWithPointer" replaceArrayWithPointer = Yes.replaceArrayWithPointer,
     Flag!"allowDownsize" allowDownsize = No.allowDownsize,
-    Range, Lengths...)(Range range, Lengths lengths)
+    Range, Lengths...)(Range range, Lengths lengths)  @trusted
```
Comment (quickfur) [inline std/experimental/ndslice/slice.d]: Why does this function need to be `@trusted`? We should be making the function that it calls `@safe` instead.

File: std/experimental/ndslice/slice.d [hunk 12]
```diff
@@ -57,7 +57,7 @@ auto sliced(
 auto sliced(
     Flag!"replaceArrayWithPointer" replaceArrayWithPointer = Yes.replaceArrayWithPointer,
     Flag!"allowDownsize" allowDownsize = No.allowDownsize,
-    size_t N, Range)(Range range, auto ref in size_t[N] lengths, size_t shift = 0)
+    size_t N, Range)(Range range, auto ref in size_t[N] lengths, size_t shift = 0) @trusted
```
Comment (quickfur) [inline std/experimental/ndslice/slice.d]: I don't like this. This function is too big and complicated to be marked `@trusted`.  Can you narrow down the parts that need to be `@trusted` and refactor them into trusted lambdas instead? Especially since this is a template function and we should be letting the compiler infer `@safe` instead of putting blind faith in the correctness of the code with `@trusted`.

File: std/experimental/ndslice/slice.d (hunk 6 above)
Comment (quickfur) [inline std/experimental/ndslice/slice.d]: That is extremely scary. What you're basically saying is that the struct is free to do anything at all and we blindly accept it as safe.

While it may be OK right now, what happens when people change the code later? They may accidentally introduce bugs that break safety, and the compiler won't even notice. That's bad.  If this is the case we may need to look into redesigning things so that we don't have to blindly trust an entire struct to do the right thing.

Basically `@trusted` should be applied to as little code as possible -- mainly the bits that the compiler cannot verify is safe even though we know better.  Everything else should be something the compiler can automatically check is `@safe`.

**Comment by quickfur** [comment]
@schveiguy I agree. That is why `@trusted` fragments of code must be as small as possible, otherwise it's impractical to verify that it's not actually doing something unsafe with stuff in its context, like affecting variables outside the trusted lambda in ways that "taint" the code outside.

If the entire function (or worse, module) is marked `@trusted`, you basically have no practical way to verify that it's actually trustworthy because there'd be too many places for untrustworthy code to hide.

Of course, just because the trusted fragment is small is no guarantee that it's actually OK. It needs to be as self-contained as possible, preferably touching as few things as possible outside of its own scope.  Because of that I'd say, any `@trusted` lambda that manipulates user-supplied template types is suspect. There is no way to guarantee the user-supplied type doesn't do all sorts of nasty things behind your back -- you have no control over what the user-supplied type does.

In the case of this module, the only obvious thing that I can think of that might require `@trusted` is pointer arithmetic involving slices. This should be very self-contained and more-or-less independent of template arguments -- so anything in this module that directly references a template type from inside a trusted lambda, I'd consider as suspect, or at the very least deserving of thorough scrutiny.

**Comment by 9il** [comment]
ndslice is used on top of raw pointers in real live, because this works faster then ndslices on top of arrays.

**Comment by schveiguy** [comment]
@9il pointer arithmetic is fine. but should be escaped with `@trusted` lambdas instead of blanket `@trusted:` blocks.

It's also OK to mark entire functions as `@trusted` if warranted, but on an individual basis, not the entire module.

**Comment by 9il** [comment]
OK, just have another more important PRs for ndslice. Will open PR for `trusted` later
